canonical/mimir-coordinator-k8s-operator
# Workers self monitoring (metrics and alert rules)

This PR adds the capability to send worker's `scrape_jobs` and alert rules to Prometheus.

It is also a continuation of the previous one: https://github.com/canonical/mimir-coordinator-k8s-operator/pull/47 and goes in tandem with https://github.com/canonical/prometheus-k8s-operator/pull/585

**TODO**: Add itest

# How to test it

1. Create a mode: `juju add-model mimir`
2. Pack and deploy mimir-coordinator: `juju deploy ./*.charm coord --resource nginx-image=ubuntu/nginx:1.18-22.04_beta --resource nginx-prometheus-exporter-image=nginx/nginx-prometheus-exporter:1.1.0 --trust`
3. Deploy a mimir-worker named `write`: `juju deploy ./*.charm write --resource mimir-image=ubuntu/mimir:2.10.0-22.04 --trust --config distributor=True --config ingester=True --config compactor=True`
4. Deploy a mimir-worker named `read`: ` juju deploy ./*.charm read --resource mimir-image=ubuntu/mimir:2.10.0-22.04 --trust --config query-frontend=True --config query-scheduler=True --config querier=True --config store-gateway=True --config compactor=True --config ruler=True --config alertmanager=True`
5. Deploy `prometheus`: `juju deploy prometheus-k8s prom --channel edge --trust `
6. Relate `write` to `coord`: `juju relate coord write`
7. Relate `coord` to `prometheus`: `juju relate prom:metrics-endpoint coord:self-metrics-endpoint`
8. Check that in prometheus we have now two more targets:
   - one for the `write` worker
   - one for the `coord` application  AKA `nginx` metrics.

    ![image](https://github.com/canonical/mimir-coordinator-k8s-operator/assets/939888/5a91462f-67a9-40bb-8ae2-c8eb800c1a2e)

   - And Alert and record rules labeled with the `write` topology 

    ![image](https://github.com/canonical/mimir-coordinator-k8s-operator/assets/939888/b7255aef-0c9c-4127-ad03-51587e7f82de)
    ![image](https://github.com/canonical/mimir-coordinator-k8s-operator/assets/939888/6b103537-a136-41f7-9dc2-30d07784ebc1)

9.  Relate `read` to `coord`: `juju relate coord read` and check you end up with this deployment:

```shell
Model  Controller  Cloud/Region        Version  SLA          Timestamp
mimir  microk8s    microk8s/localhost  3.4.0    unsupported  18:12:12-03:00

App    Version  Status   Scale  Charm                  Channel  Rev  Address         Exposed  Message
coord           waiting      1  mimir-coordinator-k8s            29  10.152.183.224  no       installing agent
prom   2.50.1   active       1  prometheus-k8s         edge     173  10.152.183.101  no       
read   2.10.0   waiting      1  mimir-worker-k8s                  0  10.152.183.212  no       installing agent
write  2.10.0   waiting      1  mimir-worker-k8s                  3  10.152.183.128  no       installing agent

Unit      Workload  Agent  Address       Ports  Message
coord/0*  blocked   idle   10.1.200.88          Incoherent deployment: you are lacking some required Mimir roles (['querier', 'query-scheduler', 'ruler', 'query-fron...
prom/0*   active    idle   10.1.200.91          
read/0*   blocked   idle   10.1.200.106         Missing mimir-cluster relation to a mimir-coordinator charm
write/0*  waiting   idle   10.1.200.121         Waiting for coordinator to publish a mimir config

Integration provider         Requirer               Interface          Type     Message
coord:self-metrics-endpoint  prom:metrics-endpoint  prometheus_scrape  regular  
coord:mimir-cluster          write:mimir-cluster    mimir_cluster      regular  
prom:prometheus-peers        prom:prometheus-peers  prometheus_peers   peer     
```



10. Check that in prometheus we only have one more target (`read`), and the are new alert rules and record rules labeled using the `read` application topology:

    ![image](https://github.com/canonical/mimir-coordinator-k8s-operator/assets/939888/5a058748-ce0e-4ec0-9668-48048eb95c28)
    ![image](https://github.com/canonical/mimir-coordinator-k8s-operator/assets/939888/c4be39d6-72be-4112-aeae-717293ccb3d7)
    ![image](https://github.com/canonical/mimir-coordinator-k8s-operator/assets/939888/a54ed309-cf45-4b1d-b75d-bfa6c03d13a7)


# Bundle:

```yaml
bundle: kubernetes
applications:
  coord:
    charm: local:mimir-coordinator-k8s-29
    scale: 1
    constraints: arch=amd64
    trust: true
  prom:
    charm: prometheus-k8s
    channel: edge
    revision: 173
    base: ubuntu@20.04/stable
    resources:
      prometheus-image: 141
    scale: 1
    constraints: arch=amd64
    storage:
      database: kubernetes,1,1024M
    trust: true
  read:
    charm: local:mimir-worker-k8s-0
    scale: 1
    options:
      alertmanager: true
      compactor: true
      querier: true
      query-frontend: true
      query-scheduler: true
      ruler: true
      store-gateway: true
    constraints: arch=amd64
    storage:
      data: kubernetes,1,1024M
      recovery-data: kubernetes,1,1024M
    trust: true
  write:
    charm: local:mimir-worker-k8s-3
    scale: 1
    options:
      compactor: true
      distributor: true
      ingester: true
    constraints: arch=amd64
    storage:
      data: kubernetes,1,1024M
      recovery-data: kubernetes,1,1024M
    trust: true
relations:
- - prom:metrics-endpoint
  - coord:self-metrics-endpoint
- - coord:mimir-cluster
  - write:mimir-cluster
```
